src/streamer.ts

```diff
@@ -0,0 +1,36 @@
+import { S3Client } from '@aws-sdk/client-s3';
+import { listBlocks, fetchStreamerMessage } from './s3fetchers';
+import { LakeConfig, BlockHeight, StreamerMessage } from './types';
+
+export async function startStream(
+  config: LakeConfig,
+  onStreamerMessageReceived: (data: StreamerMessage) => Promise<void>
+) {
+  const s3Client = new S3Client({ region: config.s3RegionName });
+
+  let lastProcessedBlockHash: string;
+  let startFromBlockHeight = config.startBlockHeight;
+  let queue: Promise<void>[] = [];
+
+  while (true) {
+    const blockHeights = await listBlocks(s3Client, config.s3BucketName, startFromBlockHeight);
+    if (blockHeights.length > 0) {
+      startFromBlockHeight = blockHeights[blockHeights.length - 1];
+    } else {
+      setTimeout(() => {}, 2000);
+      continue;
+    }
+    for (let blockHeight of blockHeights) {
+      const streamerMessage = await fetchStreamerMessage(s3Client, config.s3BucketName, blockHeight);
+      if (lastProcessedBlockHash && lastProcessedBlockHash !== streamerMessage.block.header.prevHash) {
+        startFromBlockHeight = blockHeight - 1;
+        continue;
+      }
+      queue.push(onStreamerMessageReceived(streamerMessage));
```

Comment:
Why do we need a queue? Why not just stream data to the consumer without waiting it?
Consumer application can throttle messages if it needs to.

Reply:
Don't think we want to force users to do that. cc @frol

Reply:
@luixo Do you suggest to `await` on `onStreamerMessageReceived`? There is no other way I see we could achieve throttling as streamer would run ahead without a stop and if we start from genesis it will spawn millions of `onStreamerMessageReceived` callbacks. This implementation has a pipeline that fetches the data from S3 while `onStreamerMessageReceived` is being processed, so even with a queue size of 1 there is already a benefit.

Reply:
@khorolets Let's leave a comment here explaining why the queue is there.

Reply:
We can implement async iterator: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/asyncIterator
It will let the user handle the handling speed (I kinda already forgot what's happening on the code, so I could be wrong)